hbase-balancer/src/main/java/org/apache/hadoop/hbase/master/balancer/PrefetchCacheCostFunction.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.master.balancer;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * Compute the cost of a potential cluster configuration based on the number of HFile's already
+ * cached in the bucket cache
+ */
+@InterfaceAudience.Private
+public class PrefetchCacheCostFunction extends CostFunction {
+  private static final String PREFETCH_CACHE_COST_KEY =
+    "hbase.master.balancer.stochastic.prefetchCacheCost";
+  private static final float DEFAULT_PREFETCH_COST = 500;
+
+  private String prefetchedFileListPath;
+  private double prefetchRatio;
+  private float bestPrefetchRatio;
+
+  PrefetchCacheCostFunction(Configuration conf) {
+    prefetchedFileListPath = conf.get(HConstants.PREFETCH_PERSISTENCE_PATH_KEY);
+    // Disable the prefetch cache cost function if the prefetched file list persistence is not
+    // enabled
+    this.setMultiplier(prefetchedFileListPath == null
+      ? 0.0f
+      : conf.getFloat(PREFETCH_CACHE_COST_KEY, DEFAULT_PREFETCH_COST));
+    prefetchRatio = 0.0f;
+    bestPrefetchRatio = 0.0f;
+  }
+
+  @Override
+  void prepare(BalancerClusterState cluster) {
+    super.prepare(cluster);
+    prefetchRatio = 0.0f;
+    bestPrefetchRatio = 0.0f;
+
+    for (int region = 0; region < cluster.numRegions; region++) {
+      prefetchRatio +=
+        cluster.getOrComputeWeightedPrefetchRatio(region, cluster.regionIndexToServerIndex[region]);
+      bestPrefetchRatio += cluster.getOrComputeWeightedPrefetchRatio(region,
+        cluster.getOrComputeServerWithBestPrefetchRatio()[region]);
+    }
+    prefetchRatio = bestPrefetchRatio == 0.0f ? 1.0f : prefetchRatio / bestPrefetchRatio;
```

Comment:
The prefetch ratio and the bestPrefetchRatio will be different in case the region was hosted previously on some other region server. For example, if the region was previously hosted and prefetched 100% on RS-A and later moved to RS-B and it was partially prefetched there, the best prefetch ratio will be different than the prefetch ratio on individual nodes.

Reply:
Ok, so `prefetchRation` here means the current RS hosting the region prefetch ratio.

Reply:
The bestPrefetchRatio could be 0 if a region is not yet hosted on any region server which may happen while initialising the cluster.

Reply:
Gotcha, thanks for clarifying.